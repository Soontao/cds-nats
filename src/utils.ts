import { cwdRequireCDS, User } from "cds-internal-tool";
import { headers as MsgHeaders, Msg } from "nats";
import os from "os";
import { FatalError } from "./errors";
import { HeaderObject } from "./types";


/**
 * create a timer and wait it finished
 * 
 * @param timeout 
 * @returns 
 */
export async function sleep(timeout: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, timeout));
}

/**
 * create nats headers with plain object, without context
 * 
 * @param headers 
 * @returns 
 */
export function createNatsHeaders(headers: HeaderObject = {}) {
  const msgHeaders = MsgHeaders();
  for (const [key, value] of Object.entries(headers)) {
    msgHeaders.set(key, String(value));
  }
  return msgHeaders;
}

/**
 * nats header to plain object
 * 
 * @param msg 
 * @returns plain header
 */
export function toHeader(msg: Msg): HeaderObject {
  const headers: any = {};
  if (msg.headers !== undefined) {
    for (const key of msg.headers.keys()) {
      headers[key] = msg.headers.get(key);
    }
  }
  return headers;
}

export function extractUserAndTenant(headers: HeaderObject): { user: User, id: string, tenant?: string } {
  if ("user-class" in headers && "user-attributes" in headers) {
    const cds = cwdRequireCDS() as any;
    let user;
    const tenant = headers["tenant"]; // TODO: warn if undefined
    const className = headers["user-class"];
    const id = headers["id"] ?? cds.utils.uuid();
    const userAttrs = JSON.parse(headers["user-attributes"] ?? "{}");
    if (className !== undefined) {
      if (className === "User") {
        user = new cds.User(userAttrs);
      } else {
        user = new cds.User[className](userAttrs);
      }
    }
    return { user, tenant, id };
  } else {
    throw new FatalError("fatal: user context lost from mq");
  }
}

/**
 * create nats headers with context
 * 
 * @param headers 
 * @param event 
 * @returns 
 */
export function toNatsHeaders(headers: HeaderObject = {}, event?: string) {

  const msgHeaders = createNatsHeaders(headers);

  const cds = cwdRequireCDS();

  function setIfNotExit(key: string, value: () => string) {
    if (!msgHeaders.has(key)) msgHeaders.set(key, value());
  }

  setIfNotExit("id", () => cds.context?.id ?? cds.utils.uuid());
  setIfNotExit("type", () => event ?? "unknown");
  setIfNotExit("source", () => `/default/community.cap/${os.hostname()}/${process.pid}`);
  setIfNotExit("time", () => new Date().toISOString());
  setIfNotExit("datacontenttype", () => "application/json");
  setIfNotExit("specversion", () => "1.0");
  setIfNotExit("user-class", () => {
    // @ts-ignore
    let className = cds.context?._?.user?.constructor?.name;
    if (className === undefined) className = "Anonymous";
    return className;
  });
  setIfNotExit("user-attributes", () => JSON.stringify(cds.context?.user ?? "{}"));
  if (cds.context?.tenant !== undefined) {
    setIfNotExit("tenant", () => cds.context?.tenant);
  }

  return msgHeaders;
}
