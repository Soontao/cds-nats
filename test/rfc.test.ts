/* eslint-disable @typescript-eslint/no-var-requires */
import { cwdRequireCDS, setupTest } from "cds-internal-tool";
import type NatsRFCService from "../src/NatsRFCService";
import { afterAllThings, beforeAllSetup } from "./utils";

describe("RFC Test Suite", () => {

  const cds = cwdRequireCDS();
  const axios = setupTest(__dirname, "./app");

  beforeAll(beforeAllSetup);
  afterAll(afterAllThings);

  it("should find entity metadata", async () => {
    const response = await axios.get("/people/$metadata");
    expect(response.status).toBe(200);
    expect(response.data).toMatch(/People/);
  });

  it('should support connect to RFC Service', async () => {
    const cds = cwdRequireCDS()
    const messaging = await cds.connect.to("rfc") as NatsRFCService
    expect(messaging).toBeInstanceOf(require("../src/NatsRFCService"))
    expect(messaging['appName']).toBe("demo-app-micro-service")
  });


  it('should support call remote msg with information', async () => {
    const cds = cwdRequireCDS()
    const { SELECT } = cds.ql
    const messaging = await cds.connect.to("rfc") as NatsRFCService
    const result = await messaging.execute(
      "demo-app-micro-service",
      {
        serviceName: "test.app.srv.theosun.PeopleService",
        methodName: "send",
        args: [{ query: SELECT.one.from("People").where({ Name: "Theo" }) }]
      }
    )
    expect(result).toBeNull()
  });

  it('should support shortcut of execute', async () => {
    const cds = cwdRequireCDS()
    const { SELECT, INSERT } = cds.ql
    const messaging = await cds.connect.to("rfc") as NatsRFCService
    const remoteApp = messaging.app("demo-app-micro-service");
    const rfcService = remoteApp.service("test.app.srv.theosun.PeopleService")
    const result = await rfcService.run(SELECT.one.from("People").where({ Name: "Theo" }))
    expect(result).toBeNull()
    const newName = cds.utils.uuid();
    const newPeople = await rfcService.run(INSERT.into("People").entries({ Name: newName }))
    expect(newPeople.Name).toBe(newName)
  });

  it('should support throw remote service reported error', async () => {
    const cds = cwdRequireCDS()
    const { SELECT } = cds.ql
    const messaging = await cds.connect.to("rfc") as NatsRFCService
    await expect(
      () => messaging.execute(
        "demo-app-micro-service",
        {
          serviceName: "test.app.srv.theosun.PeopleService",
          methodName: "send",
          args: [
            { query: SELECT.one.from("NotExistedEntity").where({ Name: "Theo" }) }
          ]
        },

      )
    ).rejects.toThrow("no such table")
  });

  it('should support update weight by rest api & rfc', async () => {
    let response = await axios.post("/people/People", { Name: cds.utils.uuid() })
    expect(response.status).toBe(201)
    const newPeopleId = response.data.ID
    const newPeopleName = response.data.Name
    response = await axios.post("/people/updateWeight", { ID: newPeopleId, Weight: 99.9 })
    expect(response.status).toBe(200)
    expect(response.data.Name).toBe(newPeopleName)
    const messaging = await cds.connect.to("rfc") as NatsRFCService
    const updatedData = await messaging.execute(
      "demo-app-micro-service",
      {
        serviceName: "test.app.srv.theosun.PeopleService",
        methodName: "send",
        args: [{ event: "updateWeight", data: { ID: newPeopleId, Weight: 130.2 } }]
      }
    )
    expect(updatedData).toMatchObject({ ID: newPeopleId, Name: newPeopleName, Weight: 130.2 })
  });


});
