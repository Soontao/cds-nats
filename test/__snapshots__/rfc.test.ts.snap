// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RFC Test Suite should support invoke custom handler by RFCService proxy 1`] = `[Error: method/action/function 'notExistFunction' is not existed on the app service 'test.app.srv.theosun.PeopleService' of app 'demo-app-micro-service']`;

exports[`RFC Test Suite should support throw remote service reported error 1`] = `
{
  "code": "SQLITE_ERROR",
  "errno": 1,
  "message": "SQLITE_ERROR: no such table: test_app_srv_theosun_PeopleService_NotExistedEntity in: 
SELECT * FROM test_app_srv_theosun_PeopleService_NotExistedEntity ALIAS_1 WHERE Name = ? LIMIT 1",
  "query": "SELECT * FROM test_app_srv_theosun_PeopleService_NotExistedEntity ALIAS_1 WHERE Name = ? LIMIT 1",
  "values": [
    "Theo",
  ],
}
`;
